### PR TITLE
Investigating permissive assert class

### PR DESCRIPTION
**Pull request by IvenBach**
F5 and Unit tested.

I've updated the PermissiveAssertClass to have it treat numeric comparisons as VBA does. It behaves as I expect it for numbers. Is there anything that I've missed?

Below are the tests I used to compare against:
```
'Module1
Public Function InputHasMoreThan5Characters(ByVal value As String) As Boolean
    InputHasMoreThan5Characters = Len(value) > 5
End Function

Public Function ReturningANumber() As Long
    ReturningANumber = 10
End Function
```

```
PermissiveTestModule1
Public Sub ModuleInitialize()
    'this method runs once per module.
    Set Assert = CreateObject("Rubberduck.PermissiveAssertClass")
    Set Fakes = CreateObject("Rubberduck.FakesProvider")
End Sub

'@TestMethod
Public Sub TestingShortInput()
    On Error GoTo TestFail
    
    'Arrange:

    'Act:
    Dim shortInput As Boolean
    shortInput = Module1.InputHasMoreThan5Characters("quack")
    
    'Assert:
    Assert.IsFalse shortInput

TestExit:
    Exit Sub
TestFail:
    Assert.Fail "Test raised an error: #" & Err.Number & " - " & Err.Description
End Sub

'@TestMethod
Public Sub TestingLengthyInput()
    On Error GoTo TestFail
    
    'Arrange:

    'Act:
    Dim lengthyInput As Boolean
    lengthyInput = Module1.InputHasMoreThan5Characters("More than 5 characters")
    
    'Assert:
    Assert.IsTrue lengthyInput

TestExit:
    Exit Sub
TestFail:
    Assert.Fail "Test raised an error: #" & Err.Number & " - " & Err.Description
End Sub

'@TestMethod
Public Sub TestingAreEqual()
    On Error GoTo TestFail
    
    'Arrange:
    Dim foo As Long
    foo = 10

    'Act:

    'Assert:
    Assert.AreEqual 10, CByte(10)

TestExit:
    Exit Sub
TestFail:
    Assert.Fail "Test raised an error: #" & Err.Number & " - " & Err.Description
End Sub

'@TestMethod
Public Sub TestingAreEqualStrings()
    On Error GoTo TestFail
    
    'Arrange:
    Dim bar As String
    bar = "barbarbar"
    
    'Act:

    'Assert:
    Assert.AreEqual "barbarbar", bar

TestExit:
    Exit Sub
TestFail:
    Assert.Fail "Test raised an error: #" & Err.Number & " - " & Err.Description
End Sub

'@TestMethod
Public Sub TestingSequences()
    On Error GoTo TestFail
    
    'Arrange:
    Dim expected(1 To 5) As String
    expected(1) = "Foo"
    expected(2) = "Bar"
    expected(3) = "Baz"
    expected(4) = "Duk"
    expected(5) = "End"
    
    'Act:
    Dim actual() As String
    actual = ExampleArray

    'Assert:
    Assert.SequenceEquals expected, actual

TestExit:
    Exit Sub
TestFail:
    Assert.Fail "Test raised an error: #" & Err.Number & " - " & Err.Description
End Sub

Private Function ExampleArray() As String()
    Dim temp() As String
    ReDim temp(1 To 5)
    temp(1) = "Foo"
    temp(2) = "Bar"
    temp(3) = "Baz"
    temp(4) = "Duk"
    temp(5) = "End"
    ExampleArray = temp
End Function

'@TestMethod
Public Sub TestIsNothing()
    On Error GoTo TestFail
    
    'Arrange:
    Dim foo As Worksheet
    Set foo = Sheet1
    'Act:

    'Assert:
    Assert.IsNotNothing foo

TestExit:
    Exit Sub
TestFail:
    Assert.Fail "Test raised an error: #" & Err.Number & " - " & Err.Description
End Sub

'@TestMethod
Public Sub TestingAreSame()
    On Error GoTo TestFail
    
    'Arrange:
    Dim foo As Worksheet
    Set foo = Sheet1
    
    Dim bar As Worksheet
    Set bar = Sheet2

    'Act:

    'Assert:
    Assert.AreNotSame foo, bar

TestExit:
    Exit Sub
TestFail:
    Assert.Fail "Test raised an error: #" & Err.Number & " - " & Err.Description
End Sub
```

```
StrictTestModule1

'@ModuleInitialize
Public Sub ModuleInitialize()
    'this method runs once per module.
    Set Assert = CreateObject("Rubberduck.AssertClass")
    Set Fakes = CreateObject("Rubberduck.FakesProvider")
End Sub

'@TestMethod
Public Sub TestingShortInput()
    On Error GoTo TestFail
    
    'Arrange:

    'Act:
    Dim shortInput As Boolean
    shortInput = Module1.InputHasMoreThan5Characters("quack")
    
    'Assert:
    Assert.IsFalse shortInput

TestExit:
    Exit Sub
TestFail:
    Assert.Fail "Test raised an error: #" & Err.Number & " - " & Err.Description
End Sub

'@TestMethod
Public Sub TestingLengthyInput()
    On Error GoTo TestFail
    
    'Arrange:

    'Act:
    Dim lengthyInput As Boolean
    lengthyInput = Module1.InputHasMoreThan5Characters("More than 5 characters")
    
    'Assert:
    Assert.IsTrue lengthyInput

TestExit:
    Exit Sub
TestFail:
    Assert.Fail "Test raised an error: #" & Err.Number & " - " & Err.Description
End Sub


'@TestMethod
Public Sub TestingAreEqual()
    On Error GoTo TestFail
    
    'Arrange:
    Dim foo As Long
    foo = 10

    'Act:

    'Assert:
    Assert.AreEqual CLng(10), foo

TestExit:
    Exit Sub
TestFail:
    Assert.Fail "Test raised an error: #" & Err.Number & " - " & Err.Description
End Sub

'@TestMethod
Public Sub TestingAreEqualStrings()
    On Error GoTo TestFail
    
    'Arrange:
    Dim bar As String
    bar = "barbarbar"
    
    'Act:

    'Assert:
    Assert.AreEqual "barbarbar", bar

TestExit:
    Exit Sub
TestFail:
    Assert.Fail "Test raised an error: #" & Err.Number & " - " & Err.Description
End Sub

'@TestMethod
Public Sub TestingSequences()
    On Error GoTo TestFail
    
    'Arrange:
    Dim expected(1 To 5) As String
    expected(1) = "Foo"
    expected(2) = "Bar"
    expected(3) = "Baz"
    expected(4) = "Duk"
    expected(5) = "End"
    
    'Act:
    Dim actual() As String
    actual = ExampleArray

    'Assert:
    Assert.SequenceEquals expected, actual

TestExit:
    Exit Sub
TestFail:
    Assert.Fail "Test raised an error: #" & Err.Number & " - " & Err.Description
End Sub

Private Function ExampleArray() As String()
    Dim temp() As String
    ReDim temp(1 To 5)
    temp(1) = "Foo"
    temp(2) = "Bar"
    temp(3) = "Baz"
    temp(4) = "Duk"
    temp(5) = "End"
    ExampleArray = temp
End Function

'@TestMethod
Public Sub TestIsNothing()
    On Error GoTo TestFail
    
    'Arrange:
    Dim foo As Worksheet
    Set foo = Sheet1
    'Act:

    'Assert:
    Assert.IsNotNothing foo

TestExit:
    Exit Sub
TestFail:
    Assert.Fail "Test raised an error: #" & Err.Number & " - " & Err.Description
End Sub

'@TestMethod
Public Sub TestingAreSame()
    On Error GoTo TestFail
    
    'Arrange:
    Dim foo As Worksheet
    Set foo = Sheet1
    
    Dim bar As Worksheet
    Set bar = Sheet2

    'Act:

    'Assert:
    Assert.AreNotSame foo, bar

TestExit:
    Exit Sub
TestFail:
    Assert.Fail "Test raised an error: #" & Err.Number & " - " & Err.Description
End Sub
```

### CODE REVIEW

File: RubberduckTests/UnitTesting/PermissiveAssertTests.cs [hunk 1]
```diff
@@ -0,0 +1,601 @@
...
+        {
+            var assert = new PermissiveAssertClass();
+            assert.AreNotSame(null, null);
+
+            Assert.AreEqual(TestOutcome.Failed, _args.Outcome);
+        }
+
+        //[Category("Unit Testing")]
+        //[Test]
+        //public void AreNotSameShouldFailWithSameReferences()
```
Comment (MDoerner) [inline RubberduckTests/UnitTesting/PermissiveAssertTests.cs]: Why is there a commented out test? Delete it or, if it just does not work yet, ignore it using the corresponding attribute.

File: Rubberduck.Core/UnitTesting/PermissiveObjectComparer.cs [hunk 2]
```diff
@@ -16,6 +16,12 @@ public new bool Equals(object x, object y)
             var expected = x;
             var actual = y;
 
+            //VBA will coerce any numeric-apparent-value to test them
+            if (double.TryParse(expected.ToString(), out double numericExpected) && double.TryParse(actual.ToString(), out double numericActual))
```
Comment (MDoerner) [inline Rubberduck.Core/UnitTesting/PermissiveObjectComparer.cs]: The approach to compare as doubles first is a bit naive, but I think doing it the correct way can wait until we implement the equality evaluation for the expression evaluation engine to spec. (As an example, when VBA compares a double with a single, they are compared as single.)

**Comment by ThunderFrame** [comment]
It would nice to see some more tests on arrays (right now you only test `String` arrays)
- comparing an `Empty` `Variant` to a value of `""` or `vbNullString`
- some array values having a value of `""` and/or `vbNullString`
- comparing a `Variant` array of strings to a `String` array
- comparing a `Variant` array with a mix of `Long`, `Integer`, `Byte`, `Decimal`, `Currency`, `Double`, `Single`, `Boolean` and `Date`
- Comparing a `Byte` array to a `Long` array

And some tests involving `Date`:
- ensuring that `DateSerial(2018,1,1) = CStr(DateSerial(2018,1,1))` returns `True`
- ensuring that `Clng(DateSerial(2018,1,1)) = DateSerial(2018,1,1)` returns `True`
- ensuring that `CLng(DateSerial(2018,1,1)) = CStr(DateSerial(2018,1,1)) ' raises error 13` returns `False`

And some tests involving `Boolean`:
- ensuring that `True = "true"` returns `True`
- ensuring that `"True" = "true"` returns `False`
- ensuring that `-1 = CBool("true")` returns `True`
- ensuring that `-1 = "True" ' raises error 13` returns `False`

**Comment by bclothier** [comment]
I believe this is wrong:

https://github.com/rubberduck-vba/Rubberduck/blob/d533a968cc883cc2347e16dc039913cfde9b0e80/Rubberduck.Main/ComClientLibrary/UnitTesting/AssertClass.cs#L126

In order to accurately evaluate whether the value is `Empty` or `Null`, we must be able to access the underlying information stored in the `VARIANT` structure. To do this, I think we need to modify the signature as following:

```
public void AreEqual(
  [MarshalAs(UnmanagedType.Struct)] VARIANT expected, 
  [MarshalAs(UnmanagedType.Struct)] VARIANT actual, 
  string message = ""
)
```

Or possibly as:
```
public void AreEqual(
  [MarshalAs(UnmanagedType.Struct)] IntPtr expected, 
  [MarshalAs(UnmanagedType.Struct)] IntPtr actual, 
  string message = ""
)
```

Note that `UnmanagedType.Struct` is a misnomer - it actually means to marshal as if it was a variant structure. 

Using `VARIANT` gives us direct access to the `vt` type which we can sic the `VARTYPE` enumeration on to accurately determine whether it's a `Empty` or `Null` and otherwise cast the value stored in the `data1`/`data2` to a corresponding value.

But in case this doesn't work, `IntPtr` can be used, too, which would allow use of `Marshal.PtrToStructure()` to again provide access to the VARIANT structure. 

This extra handling is necessary because with the default marshaling, the difference between a `vbNullString` and `Null` gets obfuscated. In VBA:

```
?vbNullString = Empty
True
?vbNullString = Null
Null
```

**Comment by retailcoder** [comment]
@bclothier then the `IAssert` interface needs to be changed, and the `AssertClass` implementation needs to reflect it too.

**Pull request closed without merging**